sdks/go/pkg/beam/runners/prism/internal/README.md

```diff
@@ -37,31 +39,505 @@ Handler configurations are registered by dependant packages.
+a bundle. This goroutine sends bundles on the channel when appropriate. This goroutine waits on
+a condition variable for changes from processed bundles.
+The other goroutine blocks until there are no more pending elements, at which point it
+cancels the watermark evaluating goroutine, and unblocks the condition variable, in that order.
+
+Prism's main thread when run as a stand alone command will block forever after
+initializing the beam JobManagment services. Similarly, a built in prism instance
+in the Go SDK will start a Job Management instance, and the main thread is blocked
+while the job is executing through the "Universal" runner handlers for a pipeline.
+
+A job thus has the following number of persistent goroutines in execution:
+
+For a Prism instance:
+* G for the job services.
+
+For each Job:
+* 1 for the Job Executor
+* 2 for the `ElementManager`
+
+For each Environment:
+* 2 for the Environment itself
+* 3*2 + 1 for the FnAPI Control, Data, State, and Logging Streams
+* G for the worker's GRPC server.
+
+For each Bundle:
+* 1 to handle bundle execution, up to the configured maxium parallel bundles (default 8)
+
+Letting E be the number of environments in the job, and B maximum number of parallel bundles:
+
+Total Goroutines = G + (1 + 2) + E*(3*2 +1 + 2 + G) + B(1)
+
+Total Goroutines = G + 3 + 9E + E*G + B
+
+Total Goroutines = G(E + 1) + 9E + B + 3
+
+So for a job J with 1 eviroment, and the default maximum parallel bundles, 8:
+
+Total Goroutines for Job J = G((1) + 1) + 9(1) + (8) + 3
+
+Total Goroutines for Job J = 2G + 9 + 8 + 3
+
+Total Goroutines for Job J = 2G + 20
+
+2 GRPC servers + 20 goroutines may sound like a lot, but most processes in a Prism instance
+are waiting for some trigger to execute. They are not busy waiting or spin looping.
+FnAPI messages are multiplexed so the expectation is the data service goroutines will
+be the busiest moving data back and forth from the SDK.
+
+A consequence of this approach is the need to take care in locking shared resources and data
+when they may be accessed by multiple goroutines. This,Iin particular, is all done in the `ElementManager`
+which has locks for each stage in order to serialze access to it's state. This state is notably
+accessed by the Bundle goroutines on persisting data back to the `ElementManager`.
+
+For best performance, we do as much work as possible in the Bundle Goroutines since they are
+what can most dynamically scale out, as bundle generation is handled by the single watermark
+evaluation thread. This may include migrating handling of data responses away from the Data
+received goroutine and into the Bundle processing goroutine, so bundles are less likely to
+block each other.
+
+As one G of goroutines is the single Job Management instance for prism itself, and the other is
+for the worker endpoint, these are no wasted either. Prism can assign uniques names to workers
```

Comment:
```suggestion
for the worker endpoint, these are not wasted either. Prism can assign uniques names to workers
```

Reply:
(not sure if this is the typo or not)

Reply:
Good catch. It is a typo for sure. Thanks for the dedication on this!